pkg/googlecloud/publisher.go

```diff
@@ -76,16 +79,44 @@ func NewPublisher(config PublisherConfig, logger watermill.LoggerAdapter) (*Publ
 		logger: logger,
 	}
 
-	ctx, cancel := context.WithTimeout(context.Background(), config.ConnectTimeout)
+	connectCtx, cancel := context.WithTimeout(context.Background(), config.ConnectTimeout)
 	defer cancel()
 
+	var grpcConnectionCtx context.Context
+	grpcConnectionCtx, pub.cancelCtxClient = context.WithCancel(context.Background())
+
 	var err error
-	pub.client, err = pubsub.NewClient(ctx, config.ProjectID, config.ClientOptions...)
-	if err != nil {
-		return nil, err
+
+	clientResultStream := make(chan clientResult)
+	go func() {
```

Comment:
@rvs-fluid-it I'm wondering if it would be possible to hide all this complexity in a separate func? Perhaps so we could re-use the same logic in subscriber as well?

Reply:
Good point. But how do we move from here? Maybe it's better that I prepare a new pull request containing fixes for the publisher as well for the subscriber. And I will pay attention to minimize duplication ... What do you think? Do you have other points that I need to take in to consideration?

Reply:
It's fine do update this PR or create a new one, whatever works for you. 🙂  I think the idea overall is solid, I would just pay attention so that the Publisher and Subscriber constructors are easy to follow, as it gets quite complex with `select` and two contexts. However, hiding this complexity in a separate function should be enough.

Reply:
Hello, @rvs-fluid-it! I'm doing some housekeeping in Watermill repositories and I want to know if you have any update on that? :wink: 

If you don't have time now to work on that please let me know, so I can take care of that. Thanks!